### ISSUE TITLE
Feature request: add support for Objects serialized by id (Jackson)

### ISSUE THREAD

**Issue by panzerj** [vojtechhabarta/typescript-generator#298]
I´ve got complex hierarchical data models where objects are references multiple times.

These references are written to JSON via the object id using the Jackson2 annotation
@JsonIdentityReference(alwaysAsId = true)

It would be very handy if the converter could respect this annotation and create a property with the id in typescript indead of the object

**Comment by panzerj** [comment]
Thanks for your really fast feedback and action.
I can confirm that your example works like expected. However my usage of the annotations is a little bit different .. it would be awesome if you could check it out.

I do not use the annotation `@JsonIdentityReference(alwaysAsId = true)` on the class itself, but on properties of this class´ type.

`@JsonIdentityReference(alwaysAsId = true)
    private ProjectComponent wheelType;`

This is because I define the objects of type `ProjectComponent` at exaclty one place  in the data model and these objects are serialized as objects and in all other places where they are referenced they are always serialized as id. So when a property is annotated with `@JsonIdentityReference(alwaysAsId = true)` it should always be converted to the id type.

**Comment by vojtechhabarta** [comment]
`@JsonIdentityReference` annotation is now handled also when used on properties.
(For example see `TestObjectD` and `Wrapper` in [ObjectAsIdTest](https://github.com/vojtechhabarta/typescript-generator/blob/master/typescript-generator-core/src/test/java/cz/habarta/typescript/generator/ObjectAsIdTest.java) - produced Typescript declaration in this case is `testObjectD1: string`.)

Remaining downside is that Jackson doesn't have something like `neverAsId` (in addition to `alwaysAsId`). So for example when `SomeClass` is annotated with `@JsonIdentityInfo` and its identification is of type `string` then references to this class in TypeScript are:
- either `SomeClass | string`
- or `string`

depending on `alwaysAsId` value. It is not possible to force Jackson to serialize this class always as object. Thus also TypeScript-generator is not able to produce just `SomeClass` reference type.

But maybe this is not such a problem for you, what do you think?

**Comment by ggruber4711** [comment]
Hello, 
I just stumbled above this very interesting thread. First of all, thanks for this great plugin, which saves a such a lot of work, as we have 200+ entities which shall be made available from typescript :-)
Due to large amount of entities, we tried to implement some custom jackson mappers, where we tried to implement some useful defaults, f.i.
- don't serialize collections
- serialize a ManyToOne reference always as a "truncated" object containing only the ID.
```
{
    "uid" : "11828828",
    "someStringProp" : "hello",
    "referencedObject" : {
            "uid" : "191991991991919"
    }
}
```
Question on this: Is it possible to somehow achieve a generation with a plugin which would result in wrapping the referenced object class in a reference?

```
interface MyClass {
   uid : string,
   someStringProp : string,
   referencedObject : ObjectRef<ReferencedObjectClass>
}
```
There is some plugin-architecture, but would there be a way to get something like the above implemented?

**Comment by vojtechhabarta** [comment]
Released version [2.11.472](https://github.com/vojtechhabarta/typescript-generator/releases/tag/v2.11.472) contains requested feature for serializing objects as references using IDs.

@ggruber4711 I don't think this feature can help in your case. Generally typescript-generator cannot understand what custom serializers are producing. But I think you could create an extension that would modify generated output to your needs. To create an extension you need to derive from `Extension` class and create `ModelTransformer` which transforms `TsModel`. There you can for example replace properties types (e.g. `ReferencedObjectClass` to `ObjectRef<ReferencedObjectClass>`). I think all informations needed for this should be available. As an example see `ClassEnumExtension`. Please open separate issue if you need some help or something is missing in model.

**Comment by panzerj** [comment]
Thanks for implementing this feature so quickly. It works great - at least for single objects. Would it be possible to extend it for collections also?

Using my example from above:

Generates my id string properly:
@JsonIdentityReference(alwaysAsId = true) private ProjectComponent wheelType; 

Generates ProjectComponent[] instead an array of ids
@JsonIdentityReference(alwaysAsId = true) private List<ProjectComponent> wheelTypes;
